File: inference/models/baichuan.cc
/* Copyright 2023 CMU, Facebook, LANL, MIT, NVIDIA, and Stanford (alphabetical)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "baichuan.h"       

namespace FlexFlow {

using namespace Legion;
using json = nlohmann::json;

void BAICHUAN::create_baichuan_model(FFModel &ff,
                                     std::string const &model_config_file_path,
                                     std::string const &weight_file_path,
                                     InferenceMode mode,
                                     GenerationConfig generationConfig,
                                     bool use_full_precision ) {
   BAICHUANConfig baichuan_config(model_config_file_path);
   baichuan_config.print();

  if (ff.config.tensor_parallelism_degree > baichuan_config.num_attention_heads ||
      baichuan_config.num_attention_heads % ff.config.tensor_parallelism_degree !=
          0) {
    assert(false && "The number of attention heads is smaller, or it is not "
                    "divisible by the tensor parallelism degree");
  }

  std::unordered_map<std::string, Layer *> weights_layers;   

   Tensor input;
  {
    int const token_dims[] = {BatchConfig::max_tokens_per_batch(), 1};
    input = ff.create_tensor<2>(token_dims, DT_INT32);
  }

  Initializer *embed_init = new UniformInitializer(std::rand(), 0, 0);

  Tensor token;

 //Note: the embedding layer may change 
  if (use_full_precision) {
    token = ff.embedding(input,
                         baichuan_config.vocab_size,
                         baichuan_config.hidden_size,
                         AGGR_MODE_NONE,
                         DT_FLOAT,
                         NULL,
                         embed_init,
                         "token_embedding");
  } else {
    token = ff.embedding(input,
                         baichuan_config.vocab_size,
                         baichuan_config.hidden_size,
                         AGGR_MODE_NONE,
                         DT_HALF,
                         NULL,
                         embed_init,
                         "token_embedding");
  }

  Layer *embedding = ff.layers.back();
  weights_layers.emplace("token_embedding_weight", embedding);
  Tensor w2 = nullptr;
  std::cout<<"baichuan_config.num_hidden_layers:"<<baichuan_config.num_hidden_layers<<" and use_full_precision:"<<use_full_precision <<std::endl;
  for(int i = 0; i < baichuan_config.num_hidden_layers; i++) {
    //set transformer layer id
    ff.set_transformer_layer_id(i);
    Tensor rms_norm = nullptr;
    Tensor token_rms_norm[2] = {nullptr, nullptr};
    if(i == 0) {
      rms_norm = ff.rms_norm(
        token,
        baichuan_config.rms_norm_eps,
        baichuan_config.hidden_size,
        DT_NONE, 
        std::string("layers_" + std::to_string(i) + "_attention_norm").c_str());
    } else {
      ff.residual_rms_norm(
        token,
        w2,
        token_rms_norm,
        baichuan_config.rms_norm_eps,
        baichuan_config.hidden_size,
        DT_NONE, 
        std::string("layers_" + std::to_string(i) + "_attention_norm").c_str());
        token = token_rms_norm[0];
        rms_norm = token_rms_norm[1];
    }


    //step 2: self attention
    Tensor mha;
    switch (mode) {
      case BEAM_SEARCH_MODE: {
        mha = ff.spec_inc_multihead_self_attention(
            rms_norm,
            baichuan_config.hidden_size,
            baichuan_config.num_attention_heads,
            baichuan_config.hidden_size / baichuan_config.num_attention_heads,
            baichuan_config.hidden_size / baichuan_config.num_attention_heads,
            0.0f,    /*dropout*/
            false,   /*qkv_bias*/
            false,   /*final_bias*/
            false,   /*add_zero_attn*/
            DT_NONE, /*data_type*/
            nullptr,    /*kernel_initializer*/
            true,    /*apply_rotary_embedding*/
            false,   /*scaling query*/
            1.0f,    /*scaling factor*/
            true,    /*qk_prod_scaling*/
            false,   /*position_bias*/
            std::string("layers_" + std::to_string(i) + "_attention").c_str());
        break;
      }
      case TREE_VERIFY_MODE: {
        mha = ff.inc_multihead_self_attention_verify(
            rms_norm,
            baichuan_config.hidden_size,
            baichuan_config.num_attention_heads,
            baichuan_config.hidden_size / baichuan_config.num_attention_heads,
            baichuan_config.hidden_size / baichuan_config.num_attention_heads,
            0.0f,    /*dropout*/
            false,   /*bias*/
            false,   /*add_bias_kv*/
            false,   /*add_zero_attn*/
            DT_NONE, /*data_type*/
            nullptr, /*kernel_initializer*/
            true,     /*apply_rotary_embedding*/
            false,   /*scaling query*/
            1.0f,    /*scaling factor*/
            true,    /*qk_prod_scaling*/
            false,   /*position_bias*/
            std::string("layers_" + std::to_string(i) + "_attention").c_str());
        break;
      }
      case INC_DECODING_MODE: {
        std::cout<<"ff.inc_multihead_self_attention"<<std::endl;
        mha = ff.inc_multihead_self_attention(
            rms_norm,
            baichuan_config.hidden_size,
            baichuan_config.num_attention_heads,
            baichuan_config.hidden_size / baichuan_config.num_attention_heads,
            baichuan_config.hidden_size / baichuan_config.num_attention_heads,
            0.0f,    /*dropout*/
            false,   /*bias*/
            false,   /*add_bias_kv*/
            false,   /*add_zero_attn*/
            DT_NONE, /*data_type*/
            nullptr, /*kernel_initializer*/
            true ,    /*apply_rotary_embedding*/
            false,   /*scaling query*/
            1.0f,    /*scaling factor*/
            true,    /*qk_prod_scaling*/
            false,   /*position_bias*/
            std::string("layers_" + std::to_string(i) + "_attention").c_str());
        break;
      }
      default: {
        assert(false);
      }
    }

    Layer *attention_layer = ff.layers.back();
    weights_layers.emplace("layers_" + std::to_string(i) + "_attention_weight",
                           attention_layer);

    std::string layer_name = "layers_" + std::to_string(i) + "_ffn_norm";

    Tensor token_ff_norm[2] = {nullptr, nullptr};
    ff.residual_rms_norm(token,
                         mha,
                         token_ff_norm,
                         baichuan_config.rms_norm_eps,
                         baichuan_config.hidden_size,
                         DT_NONE,
                         layer_name.c_str());

    token = token_ff_norm[0];
    Tensor ff_norm = token_ff_norm[1];

    layer_name = "layers_" + std::to_string(i) + "_feed_forward_w1";
    Tensor w1 = ff.dense(ff_norm,
                         baichuan_config.intermediate_size,
                         AC_MODE_NONE,
                         false,
                         DT_NONE,
                         nullptr,
                         nullptr,
                         nullptr,
                         REG_MODE_NONE,
                         0.0f,
                         layer_name.c_str());

    layer_name = "layers_" + std::to_string(i) + "_feed_forward_w3";
    Tensor w3 = ff.dense(ff_norm,
                         baichuan_config.intermediate_size,
                         AC_MODE_NONE,
                         false,
                         DT_NONE,
                         nullptr,
                         nullptr,
                         nullptr,
                         REG_MODE_NONE,
                         0.0f,
                         layer_name.c_str());

    Tensor multi = ff.sigmoid_silu_multi(w1, w3);

    layer_name = "layers_" + std::to_string(i) + "_feed_forward_w2";
    w2 = ff.dense(multi,
                         baichuan_config.hidden_size,
                         AC_MODE_NONE,
                         false,
                         DT_NONE,
                         nullptr,
                         nullptr,
                         nullptr,
                         REG_MODE_NONE,
                         0.0f,
                         layer_name.c_str());

    token = ff.add(token, w2);
  }

  // final normalization and linear
  Tensor final_rms_norm_output[2] = {nullptr, nullptr};
  ff.residual_rms_norm(token,
                       w2,
                       final_rms_norm_output,
                       baichuan_config.rms_norm_eps,
                       baichuan_config.hidden_size,
                       DT_NONE,
                       "norm");

  Tensor dense = ff.dense(final_rms_norm_output[1],
                          baichuan_config.vocab_size,
                          AC_MODE_NONE,
                          false,
                          DT_NONE,
                          nullptr,
                          nullptr,
                          nullptr,
                          REG_MODE_NONE,
                          0.0f,
                          "output");

  Tensor output;

  if (mode == BEAM_SEARCH_MODE) {
    Tensor softmax = ff.softmax(dense, -1);
    // output = ff.beam_top_k(softmax, baichuan_config.max_beam_width, false);
    output = ff.argmax(softmax, /*beam_Search*/ true);
  } else {
    // Tensor softmax = ff.softmax(dense, -1);
    if (generationConfig.do_sample) {
      dense = ff.scalar_truediv(dense, generationConfig.temperature, false);
      Tensor softmax = ff.softmax(dense, -1);
      output = ff.sampling(softmax, generationConfig.topp);
    } else {
      // output = ff.arg_top_k(dense, /*k=*/1, false);
      output = ff.argmax(dense, /*beam_Search*/ false);
    }
  }

    // Compile the model
  std::cout << "------start compile ----------" << std::endl;
  InferenceManager *im = InferenceManager::get_inference_manager();
  im->compile_model_and_allocate_buffer(&ff);
  FileDataLoader fileloader("",
                            weight_file_path,
                            baichuan_config.num_attention_heads,
                            baichuan_config.num_attention_heads,
                            baichuan_config.hidden_size,
                            baichuan_config.hidden_size /
                                baichuan_config.num_attention_heads,
                            ff.config.tensor_parallelism_degree);
  fileloader.load_weights(&ff, use_full_precision);
  std::cout << "------load weight finished----------" << std::endl;

  // init operators
  im->init_operators_inference(&ff);

}

} // namespace FlexFlow

File: python/flexflow/serve/models/baichuan.py
# Copyright 2023 CMU, Facebook, LANL, MIT, NVIDIA, and Stanford (alphabetical)
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from flexflow.core import *
from .base import FlexFlowModel
import random, torch, shutil

class BAICHUANConfig:
    def __init__(self, hf_config):
        print(f"hf_config: {hf_config}")
        self.max_beam_width = 1
        self.max_beam_depth = 8 
        self.num_hidden_layers = hf_config.num_hidden_layers
        self.vocab_size = hf_config.vocab_size
        self.hidden_size = hf_config.hidden_size
        self.rms_norm_eps = hf_config.rms_norm_eps
        self.intermediate_size = hf_config.intermediate_size
        self.num_attention_heads = hf_config.num_attention_heads
        self.max_position_embeddings = hf_config.max_position_embeddings
        self.num_key_value_heads = hf_config.num_attention_heads

class FlexFlowBAICHUAN(FlexFlowModel):
    def __init__(
        self,
        mode,
        generation_config,
        ffconfig,
        hf_config,
        data_type,
        max_tokens_per_batch,
        weights_filepath="",
        tokenizer_filepath=""
    ):
        self.mode = mode 
        self.generation_config = generation_config
        self.ffconfig = ffconfig
        self.data_type = data_type
        self.baichuan_config = BAICHUANConfig(hf_config)
        self.weights_filepath = weights_filepath
        self.tokenizer_filepath = tokenizer_filepath
        self.maxint = 2 ** 31 -1

        # Sanity checks
        if self.baichuan_config.hidden_size % self.baichuan_config.num_attention_heads != 0:
            raise ValueError(
                f"Hidden size ({self.baichuan_config.hidden_size}) is not divisible by n_head ({self.baichuan_config.num_attention_heads})"
            )

        # Sanity checks
        if (
            self.baichuan_config.num_attention_heads < self.ffconfig.tensor_parallelism_degree
            or self.baichuan_config.num_attention_heads % self.ffconfig.tensor_parallelism_degree != 0
        ):
            raise ValueError(
                f"Number of attention heads ({self.baichuan_config.num_attention_heads}) is smaller, or not divisible by tensor parallelism degree ({self.ffconfig.tensor_parallelism_degree})"
            )
        self.build_model(max_tokens_per_batch)

    def build_model(self, max_tokens_per_batch):
        ffmodel = FFModel(self.ffconfig)

        token_dims = [max_tokens_per_batch, 1]
        input_tensor = ffmodel.create_tensor(token_dims, DataType.DT_INT32)

        embed_init = UniformInitializer(random.randint(0, self.maxint), 0, 0)

        token = ffmodel.embedding(
            input_tensor,
            self.baichuan_config.vocab_size,
            self.baichuan_config.hidden_size,
            AggrMode.AGGR_MODE_NONE,
            self.data_type,
            None,
            embed_init,
            name = "token_embedding",
        )

        for i in range(self.baichuan_config.num_hidden_layers):
            ffmodel.set_transformer_layer_id(i)

            if i == 0:
                attn_norm = ffmodel.rms_norm(
                    token,
                    self.baichuan_config.rms_norm_eps,
                    self.baichuan_config.hidden_size,
                    name = f"layers_{i}_attention_norm",
                )

            else:
                token, attn_norm = ffmodel.residual_rms_norm(
                    token,
                    w2,
                    self.baichuan_config.rms_norm_eps,
                    self.baichuan_config.hidden_size,
                name=f"layers_{i}_attention_norm",
                )     

            if self.mode == InferenceMode.BEAM_SEARCH_MODE:
                mha = ffmodel.inc_multihead_self_attention(
                    attn_norm,
                    self.baichuan_config.hidden_size,
                    self.baichuan_config.num_attention_heads,
                    self.baichuan_config.hidden_size // self.baichuan_config.num_attention_heads,
                    self.baichuan_config.hidden_size // self.baichuan_config.num_attention_heads,   
                    0.0, #dropout
                    False, #bias 
                    False, #add_bias_kv
                    False, #add_zero_attn
                    DataType.DT_NONE, #data_type
                    None, #kernel_initializer
                    True, #apply_rotary_pos_emb
                    name=f"layers_{i}_attention",
                )
            elif self.mode == InferenceMode.TREE_VERIFY_MODE:
                mha = ffmodel.inc_multihead_self_attention_verify(
                    attn_norm,
                    self.baichuan_config.hidden_size,
                    self.baichuan_config.num_attention_heads,
                    self.baichuan_config.hidden_size // self.baichuan_config.num_attention_heads,
                    self.baichuan_config.hidden_size // self.baichuan_config.num_attention_heads,
                    0.0, #dropout
                    False, #bias
                    False, #add_bias_kv
                    False, #add_zero_attn
                    DataType.DT_NONE, #data_type
                    None, #kernel_initializer
                    True, #apply_rotary_pos_emb
                    name=f"layers_{i}_attention",
                )
            elif self.mode == InferenceMode.INC_DECODING_MODE:
                mha = ffmodel.inc_multihead_self_attention(
                    attn_norm,
                    self.baichuan_config.hidden_size,
                    self.baichuan_config.num_attention_heads,
                    self.baichuan_config.hidden_size // self.baichuan_config.num_attention_heads,
                    self.baichuan_config.hidden_size // self.baichuan_config.num_attention_heads,
                    0.0, #dropout
                    False, #bias
                    False, #add_bias_kv
                    False, #add_zero_attn
                    DataType.DT_NONE, #data_type
                    None, #kernel_initializer
                    True, #apply_rotary_pos_emb
                    name=f"layers_{i}_attention",
                )
            else:
                assert False, "Invalid mode"
            token, ff_norm = ffmodel.residual_rms_norm(
                token,
                mha,
                self.baichuan_config.rms_norm_eps,
                self.baichuan_config.hidden_size,
                name=f"layers_{i}_ffn_norm",
            )

            w1 = ffmodel.dense(
                ff_norm,
                self.baichuan_config.intermediate_size,
                ActiMode.AC_MODE_NONE,
                False,
                name=f"layers_{i}_feed_forward_w1",
            )

            w3 = ffmodel.dense(
                ff_norm,
                self.baichuan_config.intermediate_size,
                ActiMode.AC_MODE_NONE,
                False ,
                name=f"layers_{i}_feed_forward_w3",
            )

            multi = ffmodel.sigmoid_silu_multi(w1, w3)

            w2 = ffmodel.dense(
                multi,
                self.baichuan_config.hidden_size,
                ActiMode.AC_MODE_NONE,
                False,
                name=f"layers_{i}_feed_forward_w2",
            )

        token = ffmodel.rms_norm(
            token,
            self.baichuan_config.rms_norm_eps,
            self.baichuan_config.hidden_size,
            name="norm",
        )

        dense = ffmodel.dense(
            token, 
            self.baichuan_config.vocab_size,
            ActiMode.AC_MODE_NONE,
            False,
            name="output",
        )

        if self.mode == InferenceMode.BEAM_SEARCH_MODE:
            softmax = ffmodel.softmax(dense, -1)
            #output = ffmodel.beam_top_k(softmax, self.llama_config.max_beam_width, False)
            output = ffmodel.argmax(softmax, True)
        else:
            if self.generation_config.do_sample:
                dense = ffmodel.scalar_true_divide(
                    dense, self.generation_config.temperature, False
                )
                softmax = ffmodel.softmax(dense, -1)
                output = ffmodel.sampling(softmax, self.generation_config.topp)
            else:
                #output = ffmodel.arg_top_k(dense, 1, False)
                output = ffmodel.argmax(dense, False)

        self.ffmodel = ffmodel

    def convert_hf_model(model, dst_folder):
        os.makedirs(dst_folder, exist_ok=True)
        print(f"model: {model} and dst_folder: {dst_folder}")
        for name, params in model.named_parameters():
            name = (
                name.replace(".", "_")
                .replace("self_attn", "attention")
                .replace("W_pack", "Wqkv")
                .replace("o_proj", "wo")
                .replace("mlp", "feed_forward")
                .replace("gate_proj", "w1")
                .replace("down_proj", "w2")
                .replace("up_proj", "w3")
                .replace("input_layernorm", "attention_norm")
                .replace("post_attention_layernorm", "ffn_norm")
                .replace("embed_tokens", "token_embedding")
                .replace("lm_head", "output")
                .replace("model_", "")
            )
            print(f"0 name:{name} and params.shape: {params.shape} and")
            if "attention_Wqkv" in name:
                name_q = name.replace("attention_Wqkv", "attention_wq")
                name_k = name.replace("attention_Wqkv", "attention_wk")
                name_v = name.replace("attention_Wqkv", "attention_wv")
                q, k, v = torch.split(
                    params,
                    [
                        model.config.hidden_size,
                        model.config.hidden_size,
                        model.config.hidden_size,
                    ],
                    0,
                )
                q.detach().cpu().numpy().tofile(os.path.join(dst_folder, name_q))
                k.detach().cpu().numpy().tofile(os.path.join(dst_folder, name_k))
                v.detach().cpu().numpy().tofile(os.path.join(dst_folder, name_v))
                print(f"1 name:{name}  and name_q: {name_q} and name_k: {name_k} and name_v: {name_v}")
                print(f"2 q.shape: {q.shape} and k.shape: {k.shape} and v.shape: {v.shape}")
            else:
                print(f"3 name:{name} and params.shape: {params.shape} ")
                params.detach().cpu().numpy().tofile(os.path.join(dst_folder, name))